modin/core/execution/ray/implementations/pandas_on_ray/partitioning/partition.py

```diff
@@ -94,6 +94,9 @@ def apply(self, func, *args, **kwargs):
         It does not matter if `func` is callable or an ``ray.ObjectRef``. Ray will
         handle it correctly either way. The keyword arguments are sent as a dictionary.
         """
+        if self.execution_wrapper.prefer_lazy(func):
+            return self.add_to_apply_calls(func, *args, **kwargs)
```

Comment:
It seems to me we should move this to the partition manager layer.

Reply:
Also, I think we should discuss on a meeting the introduced lazy execution in general because now we are adding more cases when Modin is lazy even if it is not enabled.

Reply:
> It seems to me we should move this to the partition manager layer.

It seems reasonable, but it will require more changes, because the apply() function is called by multiple functions in PandasDataframePartitionManager.

> Also, I think we should discuss on a meeting the introduced lazy execution in general because now we are adding more cases when Modin is lazy even if it is not enabled.

I think, the MODIN_LAZY_EXECUTION variable could be extended to have 3 values: AUTO, ON, OFF. AUTO is by default and we choose which mode to use.

Reply:
> > It seems to me we should move this to the partition manager layer.
> 
> It seems reasonable, but it will require more changes, because the apply() function is called by multiple functions in PandasDataframePartitionManager.
> 
> > Also, I think we should discuss on a meeting the introduced lazy execution in general because now we are adding more cases when Modin is lazy even if it is not enabled.
> 
> I think, the MODIN_LAZY_EXECUTION variable could be extended to have 3 values: AUTO, ON, OFF. AUTO is by default and we choose which mode to use.

I see. Maybe we could do simpler just by calling `lazy_map_partitions` in the binary operator.



> > It seems to me we should move this to the partition manager layer.
> 
> It seems reasonable, but it will require more changes, because the apply() function is called by multiple functions in PandasDataframePartitionManager.
> 
> > Also, I think we should discuss on a meeting the introduced lazy execution in general because now we are adding more cases when Modin is lazy even if it is not enabled.
> 
> I think, the MODIN_LAZY_EXECUTION variable could be extended to have 3 values: AUTO, ON, OFF. AUTO is by default and we choose which mode to use.

I agree. Let's address this in a separate PR.